File: Ice/Cpp/programs/nameserver/nameserver.cpp
/****************************************************************************
FILE          : nameserver.cpp
LAST REVISION : 2006-07-13
SUBJECT       : Implementation of the Chatter nameserver.

Note that since the name services are actually provided by a Chatter
object, there is no reason why arbitrary client programs couldn't create
their own name servers. However, to be useful, the name server must be
"well known" and that suggests a stand alone program running 24/7.

Please send comments or bug reports to

     Peter C. Chapin
     Electrical and Computer Engineering Technology
     Vermont Technical College
     Randolph Center, VT 05061
     pchapin@ecet.vtc.edu
****************************************************************************/

#include <Ice/Ice.h>
#include "Chatter.h"
#include <NameService.h>

using namespace std;
using namespace chatter;

int main(int argc, char* argv[])
{
  int status = 0;
  Ice::CommunicatorPtr ic;
  try {
    ic = Ice::initialize(argc, argv);
    Ice::ObjectAdapterPtr adapter = ic->createObjectAdapterWithEndpoints(
      "ChatterAdapter", "tcp -p 9010");
    Ice::ObjectPtr object = new NameService_impl;
    adapter->add(object, ic->stringToIdentity("NameService"));
    adapter->activate();
    ic->waitForShutdown();
  }
  catch (const Ice::Exception& e) {
    cerr << e << endl;
    status = 1;
  }
  catch (const char* msg) {
    cerr << msg << endl;
    status = 1;
  }

  // We must call ic->destroy() even if an exception is throw above.
  if (ic) {
    try {
      ic->destroy();
    }
    catch (const Ice::Exception& e) {
      cerr << e << endl;
      status = 1;
    }
  }
  return status;
}

File: Ice/Java/programs/NameServer/src/edu/vtc/chatter/NameServer.java
package edu.vtc.chatter;

import com.zeroc.Ice.*;
import java.lang.Exception;

/**
 * The main class of the Chatter name server program.
 */
public class NameServer {

    /**
     * The main method initializes Ice and creates an instance of an object that does the work
     * of the name service.
     *
     * @param args Command line arguments.
     */
    public static void main(String[] args)
    {
        int status = 0;
        Communicator ic = null;
        try {
            ic = Util.initialize(args);
            ObjectAdapter adapter =
                    ic.createObjectAdapterWithEndpoints("NameServerAdapter", "default -p 9100");
            NameServiceI object = new NameServiceI();
            adapter.add(object, Util.stringToIdentity("NameService"));
            adapter.activate();
            ic.waitForShutdown();
        }
        catch (LocalException e) {
            e.printStackTrace();
            status = 1;
        }
        catch (Exception e) {
            System.err.println(e.getMessage());
            status = 1;
        }
        if (ic != null) {
            // Clean up
            try {
                ic.destroy();
            }
            catch (Exception e) {
                System.err.println(e.getMessage());
                status = 1;
            }
        }
        System.exit(status);
    }

}
